rustworkx-core/src/centrality.rs

```diff
@@ -1143,3 +1132,414 @@ where
+    use petgraph::visit::EdgeRef;
+
+    #[test]
+    fn test_weighted_closeness_graph() {
+        let g = petgraph::graph::UnGraph::<u32, f64>::from_edges([
+            (0, 1, 1.0),
+            (1, 2, 1.0),
+            (2, 3, 1.0),
+            (3, 4, 1.0),
+            (4, 5, 1.0),
+            (5, 6, 1.0),
+        ]);
+        let classic_closeness = closeness_centrality(&g, false);
+        let weighted_closeness = newman_weighted_closeness_centrality(&g, false, |x| *x.weight());
+
+        assert_eq!(classic_closeness, weighted_closeness);
+    }
+
+    #[test]
+    fn test_the_same_as_closeness_centrality_when_weights_are_1_not_improved_digraph() {
+        let g = petgraph::graph::DiGraph::<u32, f64>::from_edges([
+            (0, 1, 1.0),
+            (1, 2, 1.0),
+            (2, 3, 1.0),
+            (3, 4, 1.0),
+            (4, 5, 1.0),
+            (5, 6, 1.0),
+        ]);
+        let classic_closeness = closeness_centrality(&g, false);
+        let weighted_closeness = newman_weighted_closeness_centrality(&g, false, |x| *x.weight());
+
+        assert_eq!(classic_closeness, weighted_closeness);
+    }
+
+    #[test]
+    fn test_the_same_as_closeness_centrality_when_weights_are_1_improved_digraph() {
+        let g = petgraph::graph::DiGraph::<u32, f64>::from_edges([
+            (0, 1, 1.0),
+            (1, 2, 1.0),
+            (2, 3, 1.0),
+            (3, 4, 1.0),
+            (4, 5, 1.0),
+            (5, 6, 1.0),
+        ]);
+        let classic_closeness = closeness_centrality(&g, true);
+        let weighted_closeness = newman_weighted_closeness_centrality(&g, true, |x| *x.weight());
+
+        assert_eq!(classic_closeness, weighted_closeness);
+    }
+
+    #[test]
+    fn test_weighted_closeness_two_connected_components_not_improved_digraph() {
+        let g = petgraph::graph::DiGraph::<u32, f64>::from_edges([
+            (0, 1, 1.0),
+            (1, 2, 0.5),
+            (2, 3, 0.25),
+            (4, 5, 1.0),
+            (5, 6, 0.5),
+            (6, 7, 0.25),
+        ]);
+        let c = newman_weighted_closeness_centrality(&g, false, |x| *x.weight());
+
+        assert_eq!(
+            [
+                Some(0.0),
+                Some(1.0),
+                Some(0.4),
+                Some(0.17647058823529413),
+                Some(0.0),
+                Some(1.0),
+                Some(0.4),
+                Some(0.17647058823529413)
+            ],
+            *c
+        );
+    }
+
+    #[test]
+    fn test_weighted_closeness_two_connected_components_improved_digraph() {
+        let g = petgraph::graph::DiGraph::<u32, f64>::from_edges([
+            (0, 1, 1.0),
+            (1, 2, 0.5),
+            (2, 3, 0.25),
+            (4, 5, 1.0),
+            (5, 6, 0.5),
+            (6, 7, 0.25),
+        ]);
+        let c = newman_weighted_closeness_centrality(&g, true, |x| *x.weight());
+
+        assert_eq!(
+            [
+                Some(0.0),
+                Some(0.14285714285714285),
```

Comment:
You probably want to compare the results using something like https://github.com/Qiskit/rustworkx/blob/df38b0a553e355a8c96fbb74e67ac3a67551c854/rustworkx-core/src/centrality.rs#L538

We should aim to have the most precision we can, but in practice minor changes to the algorithm can fudge non-significant digits of the float. Define an epsilon and check if `|a - b| <= eps`.

You might want a macro that loops over the arrays as well with `zip` as well.